agent/command/macsign.go

```diff
@@ -0,0 +1,232 @@
+	}
+
+	if err = util.ExpandValues(macSign, conf.Expansions); err != nil {
+		return errors.WithStack(err)
+	}
+
+	if !filepath.IsAbs(macSign.LocalZipFile) {
+		macSign.LocalZipFile = filepath.Join(macSign.WorkingDir, macSign.LocalZipFile)
+	}
+	if !filepath.IsAbs(macSign.OutputZipFile) {
+		macSign.OutputZipFile = filepath.Join(macSign.WorkingDir, macSign.OutputZipFile)
+	}
+	if macSign.EntitlementsFilePath != "" && !filepath.IsAbs(macSign.EntitlementsFilePath) {
+		macSign.EntitlementsFilePath = filepath.Join(macSign.WorkingDir, macSign.EntitlementsFilePath)
+	}
+	if macSign.ClientBinary != "" && !filepath.IsAbs(macSign.ClientBinary) {
+		macSign.ClientBinary = filepath.Join(macSign.WorkingDir, macSign.ClientBinary)
+	}
+
+	return nil
+}
+
+func (macSign *macSign) shouldRunForVariant(buildVariantName string) bool {
+	//If no buildvariant filter, run for everything
+	if len(macSign.BuildVariants) == 0 {
+		return true
+	}
+
+	//Only run if the buildvariant specified appears in the list.
+	return utility.StringSliceContains(macSign.BuildVariants, buildVariantName)
+}
+
+// Implementation of Execute. Expands the parameters,
+// and then execute macOS signing and/or notarization.
+func (macSign *macSign) Execute(ctx context.Context,
+	comm client.Communicator, logger client.LoggerProducer, conf *internal.TaskConfig) error {
+
+	if err := macSign.expandParams(conf); err != nil {
+		return errors.WithStack(err)
+	}
+
+	if err := createEnclosingDirectoryIfNeeded(macSign.WorkingDir); err != nil {
+		return errors.Wrap(err, "problem making working directory")
+	}
+
+	if !macSign.shouldRunForVariant(conf.BuildVariant.Name) {
+		logger.Task().Infof("Skipping macsign of local file %s for variant %s",
+			macSign.LocalZipFile, conf.BuildVariant.Name)
+		return nil
+	}
+
+	// client binary
+	if macSign.ClientBinary == "" {
+		macSign.ClientBinary = "/usr/local/bin/macnotary"
+	}
+	signMode := "sign"
+	if macSign.Notarize {
+		signMode = "notarizeAndSign"
+	}
+	args := []string{"-f", macSign.LocalZipFile,
+		"-k", macSign.KeyId,
+		"-s", macSign.Secret,
+		"-u", macSign.ServiceUrl,
+		"-m", signMode,
+		"-o", macSign.OutputZipFile,
+	}
+	if macSign.EntitlementsFilePath != "" {
+		args = append(args, "-e", macSign.EntitlementsFilePath)
+	}
+	if macSign.ArtifactType != "" {
+		args = append(args, "-t", macSign.ArtifactType)
+	}
+	if macSign.BundleId != "" {
+		args = append(args, "-b", macSign.BundleId)
+	}
+	if macSign.Verify {
+		args = append(args, "--verify")
+	}
+
+	cmd := exec.Command(macSign.ClientBinary, args...)
+
+	var exitCode int
+
+	stdout, err := cmd.CombinedOutput()
+	output := string(stdout)
+
+	if err != nil {
+		exitErr, ok := err.(*exec.ExitError)
+		if !ok {
+			return fmt.Errorf("unexpected error %s: \n%s\n%s\n%s", err, output, runtime.GOOS, runtime.GOARCH)
+		}
+
+		exitCode = exitErr.ExitCode()
+	}
+
+	if exitCode != 0 {
+		return fmt.Errorf("none zero exit code: %d: \n%s", exitCode, output)
+	}
+
+	logger.Task().Info(output)
```

Comment:
Is this output likely to be long/useful? I think it may be better to log this under debugging, or as logger.Execution() (i.e. under the agent logs)

Reply:
the output of that message will contain an information about the download url of the signed artifact(along with other useful information such as request Id of the service). This would be useful, to download the signed artifact later on or ask macos notary service owners what was the problem with the specific requestId.